### DISCUSSION TITLE
optimized numpy LincombOperator

### DISCUSSION THREAD

**Discussion by ftalbrecht** [pymor/pymor#1604] [Ideas]
@TiKeil has had the idea to speed up the asssembly of dense numpy lincomb operators by:

* storing the `operators` in a 3-tensor
* using `np.einsum` to form the linear combination once the `coefficients` are evaluated

The storage requirements are similar to an array of operators, while the speedup may be drastic. So IMHO this is something that we would like to have for all ROMs.

I would like to discuss here how to make this part of pyMOR in a pyMOR-ic way.
1. One option would be to have a `NumpyLincombOperator`, that is created in the reductors (i.e. the projected matrizes are directly stored in the 3-tensor) and that performs the `np.einsum` in the `_assemble_lincomb`. This would have several implications, in particular:
   * there would be two lincomb operators, could we share code?
   * how to deal with a linear combination of dense numpy matrices arising from adding operators?

Do you see other options, pros or cons?

**Comment by pmli** [comment]
It sounds like combining `NumpyLincombOperator` with other `Operators` might be tricky. Where would you use this (I guess when solving the ROM for many parameter values) and is it possible to use the `np.einsum` trick there without introducing a new `Operator` (e.g., adding a specialized method for solving a ROM for many parameter values)?

Also, is it worth optimizing the assembly? I would guess that solving a linear system takes significantly more time than computing a linear combination of a few matrices.

**Reply by ftalbrecht** [reply]
Well, I don't have the numbers, since @TiKeil and @sdrave considered this problem and decided it was worth to switch the assembly. But I find it highly plausible that there will be a significant gain (consider a case of a `10x10` ROM with a lhs and rhs each being a linear combination of about 30 matrices/vectors, which is not uncommon in our RB settings). So I would really like to see that feature in pyMOR, the main question is just how/where to implement it the most pyMOR-ic way.

Maybe @sdrave has an opinion?

Note: IMHO, the main difficulty lies not in performing a special `assemble_lincomb`, but to store these matrices in a 3-tensor from the start...

**Reply by sdrave** [reply]
You'd have to ask @TiKeil for concrete numbers and the specific situation. However, I remember from discussion with him that the speedup was significant for our application. Regarding implementation: I would not touch `assemble_lincomb` but rather handle this in `action_LincombOperator` of `ProjectRules`: Just check whether all of the projected child operators are `NumpyMatrixOperators` with dense matrices. If so, create an `NumpyLincombOperator` instead of a `LincombOperator`.

**Reply by ftalbrecht** [reply]
Agreed that `ProjectRules` is the appropriate place to do so! The results is still a somewhat duplicate `NumpyLincombOperator`, though. But lets see how much code can be shared there ...

**Comment by artpelling** [comment]
This is interesting! I have just read into the assembly methods and I think that `NumpyMatrixOperator._assemble_lincomb` could be improved. The first line has
```
if not all(isinstance(op, NumpyMatrixOperator) for op in operators):
    return None
```
which means that the `LincombOperator` can only be assembled if all operators are `NumpyMatrixOperators` (disregarding the other `MatchingRules`).

Let `N` be a `NumpyMatrixOperator` and `O` some other `Operator` with its own `_assemble_lincomb` method, then currently the assembly of `O+N+N` would do nothing. It would be nice, if this could assemble to `O+2N`. I have just made a PR #1618 for the assembly of `ConcatenationOperator`, where a similar behaviour would be very nice to have.

What do you think about this and could somebody explain the reason why there are `MatchingRules` for some operators, whereas others have the class method?

**Reply by sdrave** [reply]
`_assemble_lincomb` is meant for backend specific code, i.e., assemble a linear combination of Eigen or Petsc matrices. Everything else is supposed to go into `AssembleLincombRules` which certainly could be made smarter. Note that the question of finding the 'optimal' operator which represents the linear combination can get arbitrarily complex ...

**Reply by artpelling** [reply]
> `_assemble_lincomb` is meant for backend specific code, i.e., assemble a linear combination of Eigen or Petsc matrices. Everything else is supposed to go into `AssembleLincombRules` which certainly could be made smarter.
I see. But why `NumpyMatrixOperator` specifically is handled in `_assemble_lincomb`, I still don't understand...

> Note that the question of finding the 'optimal' operator which represents the linear combination can get arbitrarily complex ...

Yes that is true.

Thanks for pointing out the `contract` method. This is does exactly what I want. Could you point out the differences between the `simplify` module and the `assemble` method? I guess that `assemble` is aimed at parametric operators? They still kind of do very similar things.

**Reply by sdrave** [reply]
> Could you point out the differences between the simplify module and the assemble method? I guess that assemble is aimed at parametric operators? 

Yes. You can call `contract` also on a parametric operator. For `assemble` you would need to provide a parameter. For instance, `LincombOperator` would evaluate parameter-dependent coefficients and then be able to really form the linear combination.

> They still kind of do very similar things.

Yes. At some point one should probably discuss the situation again. For instance, one could imagine a single method, which optionally gets values for the parameters.

**Reply by sdrave** [reply]
> I see. But why NumpyMatrixOperator specifically is handled in _assemble_lincomb, I still don't understand...

Numpy is treated here like any other backend. `NumpyMatrixOperator` is in this sense no different than, e.g., `FenicsMatrixOperator`.

**Reply by artpelling** [reply]
Ok, I got it now. Thanks :)
